File: packages/ios/ContentfulOptimization/Sources/ContentfulOptimization/Tracking/ScreenTrackingState.swift
struct ScreenTrackingState {
    private var lastAcceptedScreenName: String?
    private var inFlightScreenName: String?

    func shouldTrack(_ screenName: String, trackingAllowed: Bool) -> Bool {
        trackingAllowed &&
            lastAcceptedScreenName != screenName &&
            inFlightScreenName != screenName
    }

    mutating func markInFlight(_ screenName: String) {
        inFlightScreenName = screenName
    }

    mutating func markAccepted(_ screenName: String) {
        lastAcceptedScreenName = screenName
    }

    mutating func clearInFlight(_ screenName: String) {
        if inFlightScreenName == screenName {
            inFlightScreenName = nil
        }
    }

    mutating func reset() {
        lastAcceptedScreenName = nil
        inFlightScreenName = nil
    }
}

File: packages/android/ContentfulOptimization/src/main/kotlin/com/contentful/optimization/tracking/ScreenTrackingState.kt
package com.contentful.optimization.tracking

internal class ScreenTrackingState {
    private var lastAcceptedScreenName: String? = null
    private var inFlightScreenName: String? = null

    fun shouldTrack(screenName: String, trackingAllowed: Boolean): Boolean {
        return trackingAllowed &&
            lastAcceptedScreenName != screenName &&
            inFlightScreenName != screenName
    }

    fun markInFlight(screenName: String) {
        inFlightScreenName = screenName
    }

    fun markAccepted(screenName: String) {
        lastAcceptedScreenName = screenName
    }

    fun clearInFlight(screenName: String) {
        if (inFlightScreenName == screenName) {
            inFlightScreenName = null
        }
    }

    fun reset() {
        lastAcceptedScreenName = null
        inFlightScreenName = null
    }
}
